kratos/utilities/parallel_utilities.cpp

```diff
@@ -68,6 +68,33 @@ int ParallelUtilities::GetNumProcs()
 #endif
 }
 
+bool ParallelUtilities::IsInParallel()
+{
+#ifdef KRATOS_SMP_OPENMP
+    return omp_in_parallel();
+
+#elif defined(KRATOS_SMP_CXX11)
+    // This does not work with OpenMP as
+    // it also uses the main thread for
+    // execution of parallel regions
+    if (!mspMainThreadId) {
+        LockObject lock;
```

Comment:
why do we need to lock? it kind of loses the grace if in order to know if we are in a parallel section we make it serial

Reply:
ping

Reply:
check closely, this is necessary for the initialization of the Singleton and will be only used once

Anyway there is still a design issue with this particular case:
If this is the first time accessed from another thread this one will be wrongly recognized as the main thread.
I need to improve the initialization